### PR TITLE
Listener: respect the connection balancer of the redirected listener 

### LINKED ISSUE #15146: listener: performance degradation when exact balance used with original dst
*Title*: *listener:performance degradation when exact balance used with original dst*

*Description*:

We use Envoy as a sidecar, all out bound traffic is first redirected to 127.0.0.1:15001(Envoy) by iptables, then forwarded to different listeners by original dst. When `exact balance` is enabled(on all listeners), we found the connection balance being worse. Tested with 32 downstream connections, 2 workers, connection distribution between the two handlers is always 1:31(sometimes 2:30) by observing [`downstream_cx_active`](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/stats#per-handler-listener-stats).
When a new connection being received, it will be handled by `ExactConnectionBalancer`  first, and [increase `numConnections()` of the selected handler by one](https://github.com/envoyproxy/envoy/blob/v1.17.0/source/common/network/connection_balancer_impl.cc#L31). Then connection is forwarded to a new listener in `ConnectionHandlerImpl::ActiveTcpSocket::newConnection()` , and [decrease that gauge immediately](https://github.com/envoyproxy/envoy/blob/v1.17.0/source/server/connection_handler_impl.cc#L377 ).
If connections arrive quickly, there is a high chance that the first handler is always selected since `numConnections()` of all handlers are zero.

### PR DESCRIPTION
**Pull request by lambdai**
Commit Message:
If listener1 redirects the connection to listener2, the balancer field in listener2 decides whether to rebalance.
Previously we rely on the rebalancing at listener1, however, the rebalance is weak because listener1 is likely to
not own any connection and the rebalance is no-op.

Additional Description:
~~on top of #15815~~
merged main branch

Risk Level: MID. Rebalance may introduce latency. User needs to clear rebalancer field of listener2 to recover the original behavior.
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
Fix #15146 #16113
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by dmitri-d** [comment]
To see if I got the issue right:

listener1 connection count is low, it gets a connection. Then, in `newConnection()`, a new handler is picked based on the original destination address. Becase `hand_off_restored_destination_connections` and `rebalanced` are `false` and `true`, this becomes the ultimate handler of the connection, regardless of the number of connections it's already handling.

If the above is accurate, then your changes lgtm, but I'll need to take a better look at tests.

**Comment by lambdai** [comment]
> listener1 connection count is low, it gets a connection.

Thanks @dmitri-d! The linked #15146 has more details regarding listener1. In some cases including istio, the listener1(virtual outbound in istio) doesn't own any connection after the redirect. 

Now that the accounting is bad, it's likely all the workers of listener1 will always rebalance the connection to worker0's listener1...

File: test/server/active_tcp_listener_test.cc [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+  MOCK_METHOD(Network::BalancedConnectionHandlerOptRef, getBalancedHandlerByTag,
+              (uint64_t listener_tag));
+  MOCK_METHOD(Network::BalancedConnectionHandlerOptRef, getBalancedHandlerByAddress,
+              (const Network::Address::Instance& address));
+};
+class ActiveTcpListenerTest : public testing::Test, protected Logger::Loggable<Logger::Id::main> {
+public:
+  ActiveTcpListenerTest() {
+    EXPECT_CALL(conn_handler_, dispatcher()).WillRepeatedly(ReturnRef(dispatcher_));
+    EXPECT_CALL(conn_handler_, numConnections()).Times(testing::AnyNumber());
```
Comment (dmitri-d) [inline test/server/active_tcp_listener_test.cc:45]: Is it necessary ro specify the number of calls if it's `AnyNumber`? Or does that imply greater than zero?

**Comment by dmitri-d** [comment]
the test lgtm too.

File: test/server/active_tcp_listener_test.cc (hunk 1 above)
Comment (lambdai) [reply test/server/active_tcp_listener_test.cc:45]: My hunch is that AnyNumber is 0 or more. 
This PR has only 1 test case for now. I plan to migrate more test cases more conn_handler_impl_test here. Some test cases mock numConnections() and leave the numConnections() calls 0 times in this EXPECT_CALL.

**Comment by lambdai** [comment]
gentle ping @ggreenway

File: test/mocks/network/mocks.h [hunk 2]
```diff
@@ -441,6 +441,8 @@ class MockConnectionHandler : public ConnectionHandler {
   MOCK_METHOD(void, enableListeners, ());
   MOCK_METHOD(void, setListenerRejectFraction, (UnitFloat), (override));
   MOCK_METHOD(const std::string&, statPrefix, (), (const));
+
+  std::atomic<uint64_t> num_handler_connections_{};
```
Comment (ggreenway) [inline test/mocks/network/mocks.h:445]: Why is this atomic? I don't think any of the mocks are used in an unsynchronized threaded context.

File: test/mocks/network/mocks.h [hunk 3]
```diff
@@ -421,7 +421,7 @@ class MockListener : public Listener {
   MOCK_METHOD(void, setRejectFraction, (UnitFloat));
 };
 
-class MockConnectionHandler : public ConnectionHandler {
+class MockConnectionHandler : public virtual ConnectionHandler {
```
Comment (ggreenway) [inline test/mocks/network/mocks.h:424]: Why add `virtual`?

**Review by ggreenway — CHANGES_REQUESTED** [review]
I think this will be fine. I've never looked at the rebalancer code before, but I think I've familiarized myself with it. @mattklein123 do you want to take a quick look for sanity at the main code change and make sure this makes sense to you?

I think this would benefit from an integration test, verifying that multiple levels of balancing end up with the correct result.

Also, is there a good place to document that the exact-rebalancer should not be used in these use cases? I think it's just extra overhead without any benefit.

**Comment by mattklein123** [comment]
Yeah this seems fine to me at a high level. :+1:

File: test/server/active_tcp_listener_test.cc [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+using testing::Invoke;
+using testing::NiceMock;
+using testing::Return;
+using testing::ReturnRef;
+
+namespace Envoy {
+namespace Server {
+namespace {
+
+class MockTcpConnectionHandler : public Network::TcpConnectionHandler,
```
Comment (lambdai) [inline test/server/active_tcp_listener_test.cc:32]: MockConnectionHandler virtual inheritance from ConnectionHandler is required here. The diamond is described as below.
`MockTcpConnectionHandler` <- TcpConnectionHandler <- `ConnectionHandler`
`MockTcpConnectionHandler` <- MockConnectionHandler <- `ConnectionHandler`
                            
 We want MockTcpConnectionHandler gets the mock implementation of MockConnectionHandler for free but also specialized for TcpConnectionHandler.

File: test/mocks/network/mocks.h (hunk 2 above)
Comment (lambdai) [reply test/mocks/network/mocks.h:445]: Good point! Even the default behavior is not multi-threaded in mock object.
Will fix

File: test/mocks/network/mocks.h (hunk 3 above)
Comment (lambdai) [reply test/mocks/network/mocks.h:424]: A diamond inherence is introduced in the mock object: https://github.com/envoyproxy/envoy/pull/15842#discussion_r614232093

File: api/envoy/config/listener/v3/listener.proto [hunk 5]
```diff
@@ -247,6 +247,10 @@ message Listener {
   // The listener's connection balancer configuration, currently only applicable to TCP listeners.
   // If no configuration is specified, Envoy will not attempt to balance active connections between
   // worker threads.
+  //
+  // Consider the scenario that the listener X set :ref: `original_dst <config_listener_filters_original_dst>`,
+  // and the listeners Y1,Y2 set `bind_to_port <envoy_api_field_config.listener.v3.Listener.bind_to_port>` to
+  // false. The balance config should be cleared in listener X and be enabled in Y1 and Y2.
```
Comment (htuch) [inline api/envoy/config/listener/v3/listener.proto:253]: Can you re-express? I think the first part makes sense, but I think what you're trying to say after `?` is that the "connection balance config should be applied only to Y1 and Y2"? It's unclear what is meant by "cleared".

Comment (lambdai) [reply api/envoy/config/listener/v3/listener.proto:253]: `connection_balance_config` field needs to be cleared and to disable the balancer behavior. 

s/cleared/disabled in the comment?

Comment (lambdai) [reply api/envoy/config/listener/v3/listener.proto]: Updated in commit 799d2e1

Comment (htuch) [reply api/envoy/config/listener/v3/listener.proto]: Are you asking the user to disable/clear? Again, the flow of this paragraph is not very clear. A scenario is proposed out-of-the-blue and then a statement is made. I think we should empathize with the reader and start the top of this comment block and have something that makes sense in context.

File: test/integration/listener_lds_integration_test.cc [hunk 6]
```diff
@@ -409,5 +411,103 @@ TEST_P(ListenerIntegrationTest, MultipleLdsUpdatesSharingListenSocketFactory) {
...
+          dst_listener_config.mutable_use_original_dst()->set_value(false);
+          dst_listener_config.clear_listener_filters();
+          dst_listener_config.mutable_bind_to_port()->set_value(false);
+          dst_listener_config.set_name("balanced_target_listener");
+          dst_listener_config.mutable_connection_balance_config()->mutable_exact_balance();
+
+          // 127.0.0.2 is defined in FakeOriginalDstListenerFilter.
+          *dst_listener_config.mutable_address()->mutable_socket_address()->mutable_address() =
+              "127.0.0.2";
+          dst_listener_config.mutable_address()->mutable_socket_address()->set_port_value(80);
```
Comment (ggreenway) [inline test/integration/listener_lds_integration_test.cc:449]: Hard-coded ports cause issues in CI when multiple configurations run on the same machine. See other integration tests for examples of how to do this.

File: test/integration/listener_lds_integration_test.cc [hunk 7]
```diff
@@ -409,5 +411,103 @@ TEST_P(ListenerIntegrationTest, MultipleLdsUpdatesSharingListenSocketFactory) {
...
+          auto& dst_listener_config = *bootstrap.mutable_static_resources()->add_listeners();
+          dst_listener_config = src_listener_config;
+          dst_listener_config.mutable_use_original_dst()->set_value(false);
+          dst_listener_config.clear_listener_filters();
+          dst_listener_config.mutable_bind_to_port()->set_value(false);
+          dst_listener_config.set_name("balanced_target_listener");
+          dst_listener_config.mutable_connection_balance_config()->mutable_exact_balance();
+
+          // 127.0.0.2 is defined in FakeOriginalDstListenerFilter.
+          *dst_listener_config.mutable_address()->mutable_socket_address()->mutable_address() =
```
Comment (ggreenway) [inline test/integration/listener_lds_integration_test.cc:447]: This will fail on ipv6-only. See other tests for how to properly handle this.

**Review by ggreenway — CHANGES_REQUESTED** [review]
/wait

File: test/integration/listener_lds_integration_test.cc (hunk 6 above)
Comment (antoniovicente) [reply test/integration/listener_lds_integration_test.cc:449]: We shouldn't expect listen to work in environments where port 80 is considered a privileged port.

File: api/envoy/config/listener/v3/listener.proto [hunk 8]
```diff
@@ -247,6 +247,10 @@ message Listener {
   // The listener's connection balancer configuration, currently only applicable to TCP listeners.
   // If no configuration is specified, Envoy will not attempt to balance active connections between
   // worker threads.
+  //
+  // In the scenario that the listener X redirects all the connections to the listeners Y1 and Y2,
+  // you may want to disable the balance config in listener X to avoid the cost of balancing, and
```
[the remark this answers is not in the document]
Comment (lambdai) [reply api/envoy/config/listener/v3/listener.proto:252]: will update

File: test/integration/listener_lds_integration_test.cc (hunk 6 above)
Comment (lambdai) [reply test/integration/listener_lds_integration_test.cc:449]: Don't worry. This listener set `bind_to_port` to `false`. This listener doesn't create passive socket. And it's ipv4 ipv6 safe.
See `dst_listener_config.mutable_bind_to_port()->set_value(false);`

File: test/integration/listener_lds_integration_test.cc (hunk 7 above)
Comment (lambdai) [reply test/integration/listener_lds_integration_test.cc:447]: It won't fail. `bind_to_port=false` listener doesn't use an os socket.

File: api/envoy/config/listener/v3/listener.proto (hunk 8 above)
[the remark this answers is not in the document]
Comment (lambdai) [reply api/envoy/config/listener/v3/listener.proto]: add the anchor to `use_original_dst`

File: test/integration/listener_lds_integration_test.cc (hunk 6 above)
Comment (ggreenway) [reply test/integration/listener_lds_integration_test.cc:449]: Got it. Can you add a comment to this effect about why it's safe to use these hard-coded values in this case?

File: api/envoy/config/listener/v3/listener.proto (hunk 8 above)
[the remark this answers is not in the document]
Comment (ggreenway) [reply api/envoy/config/listener/v3/listener.proto]: The update looks good. Thanks!

File: test/integration/listener_lds_integration_test.cc (hunk 6 above)
Comment (lambdai) [reply test/integration/listener_lds_integration_test.cc]: Yeah... added comments and renamed to `virtual_listener`

**Review by ggreenway — APPROVED** [review]

**Comment by ggreenway** [comment]
@htuch can you sign off on the api change?

**Review by htuch — COMMENTED** [review]
/lgtm api

**Pull request merged**